### PR TITLE
Test split library side

### PR DESCRIPTION
**Pull request by mmulholla**
The library side of the existing tests after moving common code to the api repo. No new tests, just re-org of the code to work with the split. Fully tested, but may occassionaly fail due to a bug in the library code which is fixed in https://github.com/devfile/library/pull/66

### CODE REVIEW

File: tests/v2/utils/library/test_utils.go [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+// verify verifies the library contents of the specified devfile with the expected content
+func verify(devfile *commonUtils.TestDevfile) error {
+
+	commonUtils.LogInfoMessage(fmt.Sprintf("Verify %s : ", devfile.FileName))
+
+	var errorString []string
+
+	libraryData := devfile.Follower.(DevfileFollower).LibraryData
+	commonUtils.LogInfoMessage(fmt.Sprintf("Get commands %s : ", devfile.FileName))
+	commands, _ := libraryData.GetCommands(common.DevfileOptions{})
```
Comment (maysunfaisal) [inline tests/v2/utils/library/test_utils.go:260]: we should not ignore err, if the underlying func is returnin an error, then test should err out. Same for the other calls in this func

File: tests/v2/utils/library/test_utils.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+}
+
+// editCommands modifies random attributes for each of the commands in the devfile.
+func editCommands(devfile *commonUtils.TestDevfile) error {
+
+	commonUtils.LogInfoMessage(fmt.Sprintf("Edit %s : ", devfile.FileName))
+
+	var err error
+	commonUtils.LogInfoMessage(fmt.Sprintf(" -> Get commands %s : ", devfile.FileName))
+	commands, _ := devfile.Follower.(DevfileFollower).LibraryData.GetCommands(common.DevfileOptions{})
```
Comment (maysunfaisal) [inline tests/v2/utils/library/test_utils.go:296]: should not ignore err, same for `editComponents`

File: tests/v2/utils/library/test_utils.go [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+	return err
+}
+
+// editComponents modifies random attributes for each of the components in the devfile.
+func editComponents(devfile *commonUtils.TestDevfile) error {
+
+	commonUtils.LogInfoMessage(fmt.Sprintf("Edit %s : ", devfile.FileName))
+
+	var err error
+	commonUtils.LogInfoMessage(fmt.Sprintf(" -> Get commands %s : ", devfile.FileName))
```
Comment (maysunfaisal) [inline tests/v2/utils/library/test_utils.go:313]: wrong log msg

File: tests/v2/utils/library/test_utils.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+	} else {
+		commonUtils.LogInfoMessage(fmt.Sprintf("No command found in %s : ", devfile.FileName))
+	}
+
+	commonUtils.LogInfoMessage(fmt.Sprintf("Get components %s : ", devfile.FileName))
+	components, _ := libraryData.GetComponents(common.DevfileOptions{})
+	if components != nil && len(components) > 0 {
+		err := VerifyComponents(devfile, components)
+		if err != nil {
+			errorString = append(errorString, commonUtils.LogErrorMessage(fmt.Sprintf("Verfify Commands %s : %v", devfile.FileName, err)))
```
Comment (maysunfaisal) [inline tests/v2/utils/library/test_utils.go:275]: wrong log msg

**Comment by maysunfaisal** [comment]
Please rebase since https://github.com/devfile/library/pull/66 is merged to ensure there is no intermittent failure on concurrent threads.

File: tests/README.md [hunk 5]
```diff
@@ -31,198 +31,37 @@ The tests which generate devfiles with random content at run time currently cove
...
-        1. test-utils.Verify
-            1. command-test-utils.VerifyCommands
-
-            
+## Test structure
+
+* From this repository
+    - `test/v2/libraryTest/library-test.go`: The go unit test program
+    - `test/v2/utils/library/*-utils.go` : utilites, used by the test, which contain functions uniqiue to the library tests. Mostly contain the code to modify and check devfile content.
+* From the api respository
```
[the remark this answers is not in the document]
Comment (mmulholla) [reply tests/README.md]: added

File: tests/v2/utils/library/test_utils.go [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+	devfileName := testContent.FileName
+	for i := 1; i <= numDevfiles; i++ {
+
+		testContent.FileName = commonUtils.AddSuffixToFileName(devfileName, strconv.Itoa(i))
+		commonUtils.LogMessage(fmt.Sprintf("Start test for %s", testContent.FileName))
+
+		validator := DevfileValidator{}
+		follower := DevfileFollower{}
+		libraryData, err := devfileData.NewDevfileData("2.0.0")
```
[the remark this answers is not in the document]
Comment (mmulholla) [reply tests/v2/utils/library/test_utils.go:217]: The tests do not support multiple versions yet, but certainly will be added in the future.

File: tests/README.md [hunk 7]
```diff
@@ -31,198 +31,37 @@ The tests which generate devfiles with random content at run time currently cove
...
+1. Go to directory /tests/v2/libraryTest
+1. Run ```go test``` or ```go test -v```
+1. The test creates the following files:
+    1. ```./tmp/test.log``` contains log output from the tests.
+    1. ```./tmp/library_test/Test_*.yaml``` are the devfiles which are randomly generated at runtime. The file name matches the name of the test function which resulted in them being created.
+    1. If a test detects an error when comparing properties returned by the parser with expected properties
+        *  ```./tmp/library_test/Test_*_<property id>_Parser.yaml``` - property as returned by the parser
+        *  ```./tmp/library_test/Test_*_<property id>_Test.yaml``` - property as expected by the test
+
+Note: each run of the test removes the existing conents of the ```./tmp``` directory
```
[the remark this answers is not in the document]
Comment (mmulholla) [reply tests/README.md:64]: oops, was sure I deleted that, problem of too many branches lol.

File: tests/v2/utils/library/test_utils.go [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+	// numDevfiles : the number of devfiles to create for each test
+	numDevfiles = 5
+	// numThreads :  Number of threads used by multi-thread tests
+	numThreads = 5
+)
+
+// DevfileValidator struct for DevfileValidator interface defined in common utils.
+type DevfileValidator struct{}
+
+// WriteAndValidate implements DevfileValidator interface.
```
[the remark this answers is not in the document]
Comment (mmulholla) [reply tests/v2/utils/library/test_utils.go:33]: Done

File: tests/v2/utils/library/test_utils.go [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+		if err != nil {
+			commonUtils.LogErrorMessage(fmt.Sprintf("Error vaidating file : %s : %v", devfile.FileName, err))
+		} else {
+			err = verify(devfile)
+		}
+	}
+	return err
+}
+
+// DevfileFollower struct for DevfileFollower interface defined in common utils.
```
[the remark this answers is not in the document]
Comment (mmulholla) [reply tests/v2/utils/library/test_utils.go]: done

File: tests/v2/utils/library/test_utils.go (hunk 1 above)
Comment (mmulholla) [reply tests/v2/utils/library/test_utils.go]: Fixed

File: tests/v2/utils/library/test_utils.go (hunk 2 above)
Comment (mmulholla) [reply tests/v2/utils/library/test_utils.go]: Fixed

File: tests/v2/utils/library/test_utils.go (hunk 3 above)
Comment (mmulholla) [reply tests/v2/utils/library/test_utils.go]: Fixed

File: tests/v2/utils/library/test_utils.go (hunk 4 above)
Comment (mmulholla) [reply tests/v2/utils/library/test_utils.go]: Fixed

**Comment by mmulholla** [comment]
> Please rebase since #66 is merged to ensure there is no intermittent failure on concurrent threads.

Done, ran test a dozen times with now issue.

**Comment by mmulholla** [comment]
@maysunfaisal all comments now addressed.

**Review by maysunfaisal — APPROVED** [review]

**Pull request merged**